src/Microsoft.AspNetCore.Blazor.Server/DependencyInjection/ServerSideBlazorServiceCollectionExtensions.cs

```diff
@@ -75,101 +60,68 @@ public static IServiceCollection AddServerSideBlazor(this IServiceCollection ser
-            }
-
-            return AddServerSideBlazorCore(services, startupType, configure);
-        }
+                // Call TStartup's ConfigureServices method immediately
+                var startup = Activator.CreateInstance(startupType);
+                var wrapper = new ConventionBasedStartup(startup);
+                wrapper.ConfigureServices(services);
 
-        /// <summary>
-        /// Adds Server-Side Blazor services to the service collection.
-        /// </summary>
-        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
-        /// <param name="configure">A delegate to configure the <see cref="ServerSideBlazorOptions"/>.</param>
-        /// <typeparam name="TStartup">A Blazor startup type.</typeparam>
-        /// <returns>The <see cref="IServiceCollection"/>.</returns>
-        public static IServiceCollection AddServerSideBlazor<TStartup>(
-            this IServiceCollection services,
-            Action<ServerSideBlazorOptions> configure)
-        {
-            if (services == null)
-            {
-                throw new ArgumentNullException(nameof(services));
+                // Configure the circuit factory to call a startup action when each
+                // incoming connection is established. The startup action is "call
+                // TStartup's Configure method".
+                services.Configure<DefaultCircuitFactoryOptions>(circuitFactoryOptions =>
+                {
+                    var endpoint = BlazorHub.DefaultPath; // TODO: allow configuring this
+                    if (circuitFactoryOptions.StartupActions.ContainsKey(endpoint))
+                    {
+                        throw new InvalidOperationException(
+                            "Multiple Server Side Blazor entries are configured to use " +
+                            $"the same endpoint '{endpoint}'.");
+                    }
+
+                    circuitFactoryOptions.StartupActions.Add(endpoint, builder =>
+                    {
+                        wrapper.Configure(builder, builder.Services);
+                    });
+                });
             }
 
-            return AddServerSideBlazorCore(services, typeof(TStartup), configure);
+            return services;
         }
 
-        private static IServiceCollection AddServerSideBlazorCore(
-            IServiceCollection services,
-            Type startupType,
-            Action<ServerSideBlazorOptions> configure)
+        private static void AddStandardServerSideBlazorServices(IServiceCollection services)
         {
+            // Here we add a bunch of services that don't vary in any way based on the
+            // user's configuration. So even if the user has multiple independent server-side
+            // Blazor entrypoints, this lot is the same and repeated registrations are a no-op.
             services.AddSingleton<CircuitFactory, DefaultCircuitFactory>();
```

Comment:
We should think about these all being `TryAdd` so that we don't register many instances. Generally we want framework code to be idempotent, and I think we expect users to call `AddBlazor<T>` multiple times for some scenarios.

Reply:
OK cool - changed these to `TryAdd...`.